README.md
# Мета-репо для движения «создай своего бота»

[![Build Status](https://travis-ci.org/umputun/rt-bot.svg?branch=master)](https://travis-ci.org/umputun/rt-bot)

## Боты для чата подкаста Радио-Т

- Каждый бот представляет из себя полностью законченный микросервис с единой точкой входа `/event`
- `/event` вызывается для любого сообщения через `POST` с телом `{text: сообщение, username: id пользователя, display_name: имя пользователя}` (json)
- Бот может на это реагировать 2мя способами:
 - вернуть 417 (Expectation Failed) если ему нечего сказать на сообщение
 - вернуть 201 и body `{text: сообщение, bot: имя/id бота}` (json). Техт может быть markdown. EOL должны быть представлены как строки `\n`. Заголовок Content-Type в ответе должен быть `application/json`.
- Бот оформляется как контейнер с прилагаемым docker-compose.yml, для запуска которого должно хватить `git clone ... && docker-compose up -d`
- Бот (внутри контейнера) слушает на порту `8080`
- Бот может вызывать внешние сервисы, если есть такая необходимость.
- Нежелательно тянуть за собой внешние зависимости, типа баз данных, редисов и прочего.
- На запрос бот должен ответить за какое-то фиксированное время. Предлагаемый максимум 5 сек.
- Бот не получит след. сообщения пока не ответит на предыдущее.
- Бот можно писать на всем, чем хотите/умеете. Однако, результат должен быть разумен по размеру контейнера и используемым ресурсам.
- Бот должен быть **полностью собираемым вашим Dockerfile**. Не надо коммитить собранные бинарники и/или jar у прочее в этом роде.
- UPD: После разговора с авторами решили добавить `GET /info` (200) с ответом `{author: name, info: blah, commands: [xx, yyy]}`
- UPD: Бот **не будет** получать свои собственные сообщения
- UPD: Для бота должен быть определен конфиг в файле [`bot-spec.yml`](ci/README.md). Он используется для автоматического тестирования бота.

### Идеи для полезных и забавных ботов
- ведение псевдо-разумной дискуссии по варианту сири/алексы/гугла
- ответы над узкие вопросы, типа "какая погода в Москве сегодня"
- функции поиска по подкасту (шоунотам и логам)
- все остальное, что можете придумать

### Зачем вам это надо?
- для развлечения
- для добавления в чат чего-то полезного
- для славы и почета :) Мы расскажем о каждом вашем боте который будет принят и поощрим авторов, как минимум добрым словом.

### Как в этом поучаствовать?
- посмотреть на то, что сделали другие
- почитать [что пишут](https://radio-t.com/p/2016/11/06/bot/)
- сделать PR со своим ботом в отдельном каталоге. Это хозяйство должно собираться и подниматься локально, через compose build + up
- в этом PR не забыть изменить `etc/nginx.conf` и `docker-compose.yml` для вашего бота
- Приложить README.md для своего бота
- Не забыть определить конфиг [`bot-spec.yml`](ci/README.md).

### Как проверить бот в реальной жизни?
- Ваш бот будет установлен на https://bot.radio-t.com
- Можно убедиться, что он зарегистрирован сделав `GET` на https://bot.radio-t.com/list
- Можно к нему обратиться, послав `POST` на https://bot.radio-t.com/api/:botname/event и это прийдет в ваш /event
- Или послать `GET` на https://bot.radio-t.com/api/:botname/info и это прийдет в ваш /info
 ```
 > http POST https://bot.radio-t.com/api/hello/event text=Привет username=user123 display_name=Вася
 HTTP/1.1 201 Created
 Connection: keep-alive
 Content-Length: 31
 Content-Type: application/json;charset=UTF-8
 Date: Thu, 10 Nov 2016 19:46:38 GMT
 Server: nginx
 Strict-Transport-Security: max-age=63072000; includeSubdomains; preload

 {
    "bot": "hello",
    "text": "Hello!"
 }
 ```
- Проверить как это работает end-to-end можно и **нужно** [в тестовом чатике](https://gitter.im/umputun/test)
- UPD: запустить весь проект можно локально через `run_local.sh` в корне проекта. Доступен он будет на [https://localhost](https://localhost).


rt-repl-bot/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-11-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-11-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rt-repl-bot/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rt-repl-bot/compare/0.1.0...0.1.1


loro-bot/bot.py
# coding: utf-8

import os
import markovify
import ujson as json

from flask import Flask, current_app, request, Response

BASEDIR = os.path.dirname(os.path.realpath(__file__))
BOTNAME = 'Loro'
MIN_COUNT = 1
MAX_COUNT = 10
RETRIES_MULTIPLIER = 5


def silence_response():
    return Response(
        response=json.dumps({'error': True, 'status': 417}),
        status=417,
        mimetype="application/json"
    )


def process_message(message):
    args = message.replace(BOTNAME.lower(), "").strip().split()
    if len(args) < 1:
        return "Если человек напишет имя, {} попробует что-нибудь ответить".format(BOTNAME)

    model_name = args[0]
    model = current_app.text_models.get(model_name)
    if not model:
        return "У {} нет такого корпуса. Но человек может добавить его".format(BOTNAME)

    try:
        phrases_count = int(args[1])
    except (ValueError, IndexError):
        phrases_count = MIN_COUNT

    if phrases_count < MIN_COUNT:
        phrases_count = MIN_COUNT

    if phrases_count > MAX_COUNT:
        phrases_count = MAX_COUNT

    max_retries = phrases_count * RETRIES_MULTIPLIER
    phrases = []
    retries = 0
    while len(phrases) < phrases_count and retries < max_retries:
        retries += 1
        # phrase = model.make_short_sentence(140)
        phrase = model.make_sentence()
        if phrase:
            phrases.append(phrase)

    result = ''
    for phrase in phrases:
        result += phrase
        if phrases.index(phrase) != len(phrases) - 1:
            result += '\n'
            result += '\n'

    if not result:
        result = "{} не может придумать ответ".format(BOTNAME)

    return result


def event_handler():
    try:
        data = json.loads(request.data)
    except ValueError:
        return silence_response()

    message = data.get('text')
    if not message:
        return silence_response()

    message = message.strip()
    message = message.lower()
    if not message.startswith(BOTNAME.lower()):
        return silence_response()

    result = process_message(message.lower())

    return Response(
        response=json.dumps({'text': result, 'bot': BOTNAME}),
        status=200,
        mimetype="application/json"
    )


def get_text_models():
    corpus_dir = os.path.join(BASEDIR, 'corpuses')
    corpus_files = [f for f in os.listdir(corpus_dir) if os.path.isfile(os.path.join(corpus_dir, f))]

    models = dict()

    for filename in corpus_files:
        with open(os.path.join(corpus_dir, filename), 'r', encoding='utf-8') as file:
            models[filename] = markovify.Text(file.read(), state_size=2)

    return models


def start_app():
    print('create_app')
    app = Flask('bot_app')
    app.text_models = get_text_models()
    app.config.DEBUG = True

    app.add_url_rule('/event', 'event', event_handler, methods=['POST',], strict_slashes=False)

    return app

if __name__ == '__main__':
    app = start_app()
    app.run()
